### ISSUE TITLE
Replacing values for unique jobs is unsafe when the job is executing?

### ISSUE THREAD

**Issue by dsdshcym** [sorentwo/oban#635]
> ### Precheck
> 
> - Do a quick search and make sure the bug has not yet been reported
> - For support, favor using the Elixir Forum, Slack, IRC, etc.
> - Be friendly and polite!
> 
> ### Environment
> 
> - Oban Version
> - PostgreSQL Version
> - Elixir & Erlang/OTP Versions (`elixir --version`)

Hi @sorentwo,
I'm planning to use the replacing values feature for unique jobs,
but I think Oban may not handle this edge case well.

### Current Behavior

1. When BasicEngine fetches a job to execute, [it would hold a row-level lock](https://github.com/sorentwo/oban/blob/785a2a787249081d26903a14dc53b888094ad0f1/lib/oban/queue/basic_engine.ex#L109) (SELECT FOR UPDATE)
2. Then BasicEngine [changes the job state to executing, and releases the row-level lock](https://github.com/sorentwo/oban/blob/785a2a787249081d26903a14dc53b888094ad0f1/lib/oban/queue/basic_engine.ex#L116-L129)
3. During the execution, [no lock is held by Oban](https://github.com/sorentwo/oban/blob/785a2a787249081d26903a14dc53b888094ad0f1/lib/oban/queue/producer.ex#L234-L242)
4. If we insert the same job with different `args` during the job execution: `UniqueWorker.new(new_args, replace: [:args])`, [the args would be updated](https://github.com/sorentwo/oban/blob/785a2a787249081d26903a14dc53b888094ad0f1/lib/oban/queue/basic_engine.ex#L293), but the executor is still running with the old args.
5. Then if the executor completes the job successfully, [the job state would change to completed](https://github.com/sorentwo/oban/blob/785a2a787249081d26903a14dc53b888094ad0f1/lib/oban/queue/basic_engine.ex#L134-L138), while the args is now a different value than the one used in executor.

### Expected Behavior

I'd expect Oban to not produce `completed jobs with args that are not used during the execution`
It seems our only option is to disallow `replace during executing`?
Maybe we can change the behaviour to either:
1. Do not replace args or other fields until job leaves executing state (i.e. do not allow replace when executing)
2. Allow specifying granular unique behaviours on conflict (e.g. replace when non-executing, do not replace when executing)

**Comment by sorentwo** [comment]
@dsdshcym That's a sneaky edge case! Nicely spotted. Of the potential fixes you've presented the first option (disallow replacement for executing jobs) is the cleanest.

We're pushing toward a v2.11.0 release right now and I won't have time to get this in before then. It may have to wait for a v2.11.1 release.

**Comment by dsdshcym** [comment]
Awesome, looking forward to the next release!

Thank you :)

**Comment by sorentwo** [comment]
@dsdshcym I've thought about this a bit before jumping into code fixes, and I've realized that fields should only be replaced on `available`, `scheduled`, or `retryable` jobs. Changing fields like `args` or `worker` _after_ a job's finished is potentially misleading.

What are your thoughts?

**Comment by dsdshcym** [comment]
@sorentwo 
Yes, changing fields like `args` when a job is completed doesn't make any sense for my use case:

> a job can and will only run once, but after it finished, the same job can be inserted again

```elixir
unique: [
  period: :infinity,
  fields: [:args, :worker],
  states: [:available, :scheduled, :executing, :retryable]
]
```

But I'm not sure if:
1. others may want to replace these fields after a job is finished/failed in their use cases (I doubt it but I think it's possible)
2. others may want to replace other fields when/after executing (I can think of some cases, like replacing `meta`, `max_attempts`)
So disabling replacing args/worker after certain states might be too restrictive?

If I were designing it, I would allow users to set different on_conflict behaviors based on the job states:
```
Worker.new(args, unique: [:available, :executing, :discarded], replace: %{executing: [], discarded: [:max_attempts], available: [:args])
```
(If the `replace` can accept an 1-arity anonymous function with the job as the input, that would be the best, but it might be hard if we want to persist the `replace` values)

**Comment by dsdshcym** [comment]
BTW, the more I use Oban, the more I wish the job records can be more "immutable", if you know what I mean

Basically what I want is some kind of linear logs in the db, so I can know what happened exactly over time, and when
(Currently, some fields would be overridden when a job gets retried and updated)

For example, if I were rewriting our current logic, I'd never allow a job to be retried, but to only insert a new job with the same `args` and everything to re-execute this logic
So that I can save more info for each run (error, attempted_at, node, etc.), and get better observability & traceability

**Comment by sorentwo** [comment]
>  others may want to replace these fields after a job is finished/failed in their use cases (I doubt it but I think it's possible)

Some people may find it restrictive, but I'm not convinced more flexibility is the right answer. I'm still of the mind that updating an `executing` or _executed_ job is a bug.

> If I were designing it, I would allow users to set different on_conflict behaviors based on the job states:

We _are_ designing it!

> The more I use Oban, the more I wish the job records can be more "immutable", if you know what I mean.

I get what you mean. Having a datomic-style purely immutable set of jobs would be nice. I've always aimed to straddle the purity of history (state timestamps, errors, attempt counts), and the practicality of mutable records. There are probably improvements to how data is recorded!

**Comment by dsdshcym** [comment]
> but I'm not convinced more flexibility is the right answer. I'm still of the mind that updating an executing or executed job is a bug.

Yeah, I agree.

In my current app, disabling `replace` for `executing`, `disgarded`, `canceled` would definitely work.

But `executing` would still be tricky.
Let me explain our use case here:
our worker's going to broadcast some updates, so we'd like the `args` to be always up-to-date.
- When the job is pre-executing, we can use `unique + replace` to update the `args`.
- When the job is post-executing, we can insert a new job (with no `unique` check for these statuses)
- But when the job is executing, 
  we don't want to update the `args` (due to reasons discussed above),
  and we don't want to insert a new job either (since the current job might get to `retryable` status and we'd have a race condition)
  Luckily, we already implemented a mechanism in our business logic to retry inserting the new `args` again later, so we can just do nothing in this case.

That's why I proposed a `replace: %{executing: [], available: [:args], ...}` option
(to not insert a new job when executing, and replace in other statuses)
But if Oban disallows replacing for `executing`, that would work for us as well.
